### ISSUE TITLE
Races involving channels can result in dropped messages

### ISSUE THREAD

**Issue by avaragado** [yelouafi/redux-saga#546]
I'm using channels extensively for private communication between sagas. I have a `watch` saga that races three channels: one for receiving requests, one for receiving responses, and one for handling details of the request itself.

It seems the race effect guarantees a single winner, even if more than one channel has a message. The messages from the losers are dropped.

Here's a sagaMonitor screenshot which I hope demonstrates the problem:

![monitor](https://cloud.githubusercontent.com/assets/886041/18523064/7df3b102-7aaa-11e6-9424-2a90d0fd2b66.png)

In the code, I have:

```js
const { request, handler, response } = yield race({
    request: take(channelRequest),
    handler: take(channelHandler),
    response: take(channelResponse),
});
```

In the case shown in the screenshot, only `handler` was set after that code. The other two variables were `undefined`. Next time `response` won the race, its message was NOT the one shown in the screenshot: it was a subsequent one.

**Comment by yelouafi** [comment]
>It seems the race effect guarantees a single winner, even if more than one channel has a message. The messages from the losers are dropped.

Effectively `race` by design will cancel all losers. Could you explain what's the expected behavior here? If you can setup a runnable gist/jsbin... that would be helpful.

**Comment by avaragado** [comment]
If I have time later or over the weekend I'll try to set something up.

The issue here is that two independent channels had messages available to take at the same time, because of buffering. Two possible expected behaviours spring to mind:

1. The result of the race is an object containing all properties with messages available to take. In my example, both `handler` and `response` would be set. It would be up to subsequent code to cope with that scenario (eg replacing `if (request) {} else if (handler) {} else if (response) {}` with three independent `if` clauses not strung together with `else`). Documentation would cover this case, eg "When you're using race with multiple independent channels, more than one channel may have a message available, so there may be more than one winner".
1. Keep a single winner, and ensure that any other takes with messages available don't drop the messages. Assuming the race occurs in a loop, next time around the loop one of those other takes would win.

I prefer the first option, as it doesn't assume the race is looping. It's an exception to the "only one winner" rule for races that's pretty easy to describe with an example like the above.

**Comment by Andarist** [comment]
I think the second option is more natural - `race` should have just one winner. I've tried to reproduce your issue with channels dropping messages, but no luck so far.

**Comment by athibaud** [comment]
ran into this too.
the second option seems more natural to me. `race` should only have one winner, agreed.

**Comment by Andarist** [comment]
R u able to post a fiddle with reproduced bug?

**Comment by athibaud** [comment]
hum, interesting.. i had a [quick go](http://jsfiddle.net/Ljjcj4b5/5/) at trying to reproduce but the bug didn't appear..

my use case was essentially that except with more deeply nested puts (forks within forks)
and i was seeing this in react-native.. i'll do my best to reproduce when i have time.

**Comment by Andarist** [comment]
@avaragado @athibaud 
Do you think it could be connected to the issue fixed by https://github.com/yelouafi/redux-saga/pull/583 ?

In general It could have been hard to reproduce cause just after `race` being resolved bug might have not been surfaced yet, just after the next tick. That is because injected value by the saga as the outcome of `race` is geting handled synchronously and therefore until another `yield` lines are geting executed as go (obviously ;) ).

In the same time if you were racing possibly synchronous effects (`take`s from `channel`s) their synchronous callbacks were queued one after another in a loop, synchronous ofc too. So in the end sequence of events were like:

1. running race effect
2. trying to resolve 1st take
3. resolved -> took action from the channel -> immediately/synchronously resolve the race with it
3. trying to resolve 2nd take
4. resolved -> took action from the channel -> race already resolved, but msg got already dropped

**Comment by athibaud** [comment]
@Andarist this seems like exactly what was happening. (also, one of the channels was 'fed' by an external/async eventChannel so it makes it slightly more likely that the double message scenario could happen).

unfortunately i have switched my implementation since running into that bug so i cannot confirm :(

**Comment by Andarist** [comment]
Fix already merged in, soon be available in next release

**Comment by mjrussell** [comment]
I think I recently ran into this, and I just want to make sure I understand it completely and the path going forward (next release):

Originally I had something like this:

```js
const { itemA, itemB } = yield race({
   itemA: take(queueA),
   itemB: take(queueB),
});

if (itemA) {
  // Do stuff with A
} else if (itemB) {
 // Do stuff with B
}
```
After reading this it seems like what is happening is that there can be a case where both take's resolve (one of the queues sources is from a socket and the other from a store dispatch) and therefore I have both `itemA` and `itemB` as non-null. And since I've got an else if, the code skips doing anything with `itemB` when `itemA` is non-null. This seems like an easy bugfix on my end, just moving to two `if` statements.

So my question is how would this look following the "fix"? Would my original code have been correct?

**Comment by Andarist** [comment]
If I understand you correctly your code is fine, however for now you might lose some buffered items WHEN in the exact moment of `yield race` you have waiting items in both queues. In the same moment they wont be both returned to the saga, only the first one will be, and the second one will get dropped in the bakcground silently without you even knowing that.

Fix is already on the master, I'll try to have it published asap as `0.12.1`.

**Comment by Andarist** [comment]
It took some time, but the fix finally has landed in the newest version! `v0.12.1` has been published.
